src/crd.rs

```diff
@@ -60,6 +62,109 @@ where
+/// specified timeout is reached (or indefinitely).
+///
+/// This is intended to be used in pre-flight checks of operators to ensure that all CRDs they
+/// require to work properly are present in Kubernetes.
+///
+/// # Parameters
+/// - `client`: Client to connect to Kubernetes API and create the CRD with.
+/// - `names`: The list of CRDs to check
+/// - `delay`: If specified, waits for the given `Duration` before checking again if all
+///     CRDs are present. If not specified defaults to 60 seconds.
+/// - `timeout`: If specified, keeps checking for the given `Duration`. If not specified,
+///     retries indefinitely.
+pub async fn wait_until_crds_present(
+    client: &Client,
+    names: Vec<&str>,
+    delay: Option<Duration>,
+    timeout: Option<Duration>,
+) -> OperatorResult<()> {
+    let delay = delay.unwrap_or_else(|| Duration::from_secs(60));
+    let start = Instant::now();
+
+    // The loop will continue running until either all CRDs are present or a configured
+    // timeout is reached
+    loop {
+        debug!(
+            "Checking if the following CRDs have been created: {:?}",
+            names
+        );
+
+        // Concurrently use `check_crd` to check if CRDs are there, this returns a Result containing
+        // a tuple (crd_name, presence_flag) which is collected into a single result we can then
+        // check
+        // If any requests to Kubernetes fail (crd missing is not considered a failure here) the
+        // remaining futures are aborted, as we wouldn't be able to use the results anyway
+        let check_result = futures::future::try_join_all(
+            names
+                .iter()
+                .map(|crd_name| check_crd(&client, crd_name))
+                .collect::<Vec<_>>(),
+        )
+        .await
+        // Any error returned here was an error when talking to Kubernetes and will mark this
+        // entire iteration as failed
+        .and_then(|crd_results| {
+            debug!("Received results for CRD presence check: {:?}", crd_results);
+            let missing_crds = crd_results
+                .iter()
+                .filter(|(_, present)| !*present)
+                .map(|(name, _)| String::from(name))
+                .collect::<HashSet<_>>();
+            if missing_crds.is_empty() {
+                Ok(())
+            } else {
+                Err(RequiredCrdsMissing {
+                    names: missing_crds,
+                })
+            }
+        });
+
+        // Checks done, now we
+        //   1. return ok(()) if all CRDs are present
+        //   2. return an error if CRDs are missing and the timeout has expired
+        //   3. queue another loop iteration if an error occurred and the timeout has not expired
+        match check_result {
+            Ok(()) => return Ok(()),
+            Err(err) => {
+                info!(
+                    "Error occurred when checking if all required CRDs are present: [{}]",
```

Comment:
If I understand correctly this will show even for missing CRDs which does not feel like an "error" to me.

It'd be good if you could distinguish between actual errors and missing CRDs.
Errors should probably be `error!` as well while missing CRDs should probably print which CRDs are still missing (which this will do)

Reply:
True, I'll split that into two different match branches. 

Regarding the log level, I've followed the principle of logging at debug and info here and leaving anything more severe up to the code using this library. Otherwise we'll probably have duplicate logging once here and then the operator will log an error again.
That being said, this might be a bit of a special case since we probably won't return control to the operator here until all CRDs are present (unless the operator specifies a timeout and adds a loop of its own) so logging at warn might make sense..